torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2164,10 +2171,26 @@ def _full_post_state_dict_hook(
         if not state_dict or not self._fsdp_wrapped_module.has_params:
             return state_dict
 
-        # If the `FlatParameter` is registered, then this rank only needed to
-        # participate in the all-gather but does not actually save the state
-        # dict (e.g. when `rank0_only=True` and `self.rank != 0`)
-        if hasattr(self._fsdp_wrapped_module, "flat_param"):
+        # If a rank has already exited the `summon_full_params()` context here
+        # (e.g. when `rank0_only=True` and `rank != 0`), then the rank only
+        # needed to participate in teh all-gather and does not need to save the
+        # state dict. For `use_orig_params=False`, we can check this via
+        # `FlatParameter` registration.
+        # TODO: For `use_orig_params=True`, we check for the reshard upon
+        # exiting `summon_full_params()` via the parameter shape. However, for
+        # `NO_SHARD`, we cannot tell from the shape, so we do not return early.
+        if (
+            (
+                not self._use_orig_params
+                and "flat_param" in self._fsdp_wrapped_module._parameters
+            )
+            or (
+                self._use_orig_params
+                and self._handles
+                and self._handles[0].uses_sharded_strategy
+                and self._handles[0].is_sharded(self._handles[0].flat_param)
```

Comment:
is `is_sharded` the canonical, recommended way to check if a param is in the sharded state? how about gradients?

In general, is it worth exposing docs on such methods to aid FSDP developers in the future who are looking to do these common sort of things?

Reply:
Yes, `is_sharded()` can be the canonical way to check if a parameter or its gradient is currently sharded.

Do you have any suggestions for how to expose docs / what would help FSDP developers onboard more efficiently? The method is currently documented, but perhaps this is not salient enough.